subprojects/security/src/main/java/org/gradle/security/internal/pgp/BaseInMemoryPgpSignatoryProvider.java

```diff
@@ -43,26 +44,24 @@ public class BaseInMemoryPgpSignatoryProvider implements BaseSignatoryProvider<P
 
     private final PgpSignatoryFactory factory = new PgpSignatoryFactory();
     private final Map<String, PgpSignatory> signatories = new LinkedHashMap<>();
-    private final String defaultKeyId;
-    private final String defaultSecretKey;
-    private final String defaultPassword;
+    private final Lazy<PgpSignatory> defaultSignatory;
 
     public BaseInMemoryPgpSignatoryProvider(String defaultSecretKey, String defaultPassword) {
         this(null, defaultSecretKey, defaultPassword);
     }
 
     public BaseInMemoryPgpSignatoryProvider(String defaultKeyId, String defaultSecretKey, String defaultPassword) {
-        this.defaultKeyId = defaultKeyId;
-        this.defaultSecretKey = defaultSecretKey;
-        this.defaultPassword = defaultPassword;
+        this.defaultSignatory = Lazy.locking().of(() -> {
+            if (defaultSecretKey != null && defaultPassword != null) {
+                return createSignatory("default", defaultKeyId, defaultSecretKey, defaultPassword);
```

Comment:
❓ Should we try to prevent mutation of `defaultSecretKey` and `defaultPassword` after the signature has been computed?

Reply:
>Should we try to prevent mutation of defaultSecretKey and defaultPassword after the signature has been computed?

I don't fully understand your point. The key and password stored in this class could not be mutated before, and cannot be mutated now. You can change the default signature but it would replace the provider in `SigningExtension`.

You can, in theory, change the signature between sign task actions and get surprising result, but I feel that addressing that possibility is out of scope of this issue.

Reply:
Nevermind, I was reading these as task fields, not as final fields of an utility class.